proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+
+#### Error type inference from `for try await` loops
+
+The `Failure` associated type is only accessible at runtime in the Swift 5.11 standard library; code running against older standard library versions does not include the `Failure` requirement in the witness tables for `AsyncSequence` and `AsyncIteratorProtocol` conformances. This impacts error type inference from `for try await` loops.
+
+When the thrown error type of an `AsyncIteratorProtocol` is available, either through the associated type witness (because the context has appropriate availability) or because the iterator type is concrete, iteration over an async sequence throws its `Failure` type:
+
+```swift
+struct MyAsyncIterator: AsyncIteratorProtocol {
+  typealias Failure = MyError
+  ...
+}
+
+func iterate<S: AsyncSequence>(over s: S) where S.AsyncIterator == MyAsyncIterator {
+  let closure = {
+    for try await element in s {
+      print(element)
+    }
+  }
+}
+```
+
+In the above code, the type of `closure` is `() async throws(MyError) -> Void`.
+
+When the thrown error type of an `AsyncIteratorProtocol` is not available, iteration over an async sequence throws `any Error`:
+
+```swift
+@available(SwiftStdlib 5.1, *)
+func iterate(over s: some AsyncSequence) {
+  let closure = {
+    for try await element in s {
+      print(element)
+    }
+  }
+}
+```
+
+In the above code, the type of `closure` is `() async throws(any Error) -> Void`.
+
+When the `Failure` type of the given async sequence is constrained to `Never`, `try` is not required in the `for-in` loop:
+
+```swift
+struct MyAsyncIterator: AsyncIteratorProtocol {
+  typealias Failure = Never
+  ...
+}
+
+func iterate<S: AsyncSequence>(over s: S) where S.AsyncIterator == MyAsyncIterator {
+  let closure = {
+    for await element in s {
+      print(element)
+    }
+  }
+}
+```
+
+In the above code, the type of `closure` is `() async -> Void`.
+
+### Adopting primary associated types
+
+The `Element` and `Failure` associated types are promoted to primary associated types. This enables using constrained existential and opaque `AsyncSequence` and `AsyncIteratorProtocol` types, e.g. `some AsyncSequence<Element, Never>` or `any AsyncSequence<Element, any Error>`.
+
+### Adopting isolated parameters
+
+The `nextElement()` requirement abstracts over actor isolation using [isolated parameters](/proposals/0313-actor-isolation-control.md). For callers to `nextElement()` that pass an iterator value that cannot be transferred across isolation boundaries under [SE-0414: Region based isolation](/proposals/0414-region-based-isolation.md), the call is only valid if it does not cross an isolation boundary. Whether `nextElement()` shares the isolation of the caller depends on the isolated argument value, with generalized rules for `nonisolated` and global-actor isolation described below.
+
+#### Generalized `isolated` parameters
```

Comment:
Just a nudge to remove this (and probably makes sense to add a cross-reference once we have a permalink in place)